### PR TITLE
issue #94

### PR DESCRIPTION
**Pull request by N00Bmaster2810**
## Issue that this pull request solves

 Closes: #94 under SWOC

### Brief description of what is fixed or changed

- Add a route for changing password. [PUT] /admin/password
- Validation Schema for the request body
- API Testing using Postman demonstrating all possible cases(valid, invalid) with test cases

## Types of changes

_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update (Documentation content changed)
- [ ] Other (please describe): 

## Checklist

_Put an `x` in the boxes that apply_

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] My changes does not break the current system and it passes all the current test cases.

## Screenshots

Please attach the screenshots of the changes made in case of change in user interface

## Other information

API testing collection link 
https://www.getpostman.com/collections/aba0b43b7e2e18f7d9a7

### CODE REVIEW

**Comment by jackfrost13** [comment]
@N00Bmaster2810 I will review it tomorrow.
Had a quick look. You also need to check that the new password cannot be same as old password.
User already exists check doesn't seem of any sense to me. Please elaborate on your login for this.

**Comment by N00Bmaster2810** [comment]
> @N00Bmaster2810 I will review it tomorrow.
> Had a quick look. You also need to check that the new password cannot be same as old password.
> User already exists check doesn't seem of any sense to me. Please elaborate on your login for this.

I have updated accordingly and an updated API testing collection link is:
https://www.getpostman.com/collections/aba0b43b7e2e18f7d9a7

**Review by Kajol-Kumari — APPROVED** [review]
LGTM :+1:

**Comment by N00Bmaster2810** [comment]
> LGTM

Thank you ma'am

**Comment by jackfrost13** [comment]
@N00Bmaster2810 I have a few minor suggestions. Reverting the PR. 
Another thing can you please modify the API testing such that the request body is a raw json instead of form. It will be helpful for integration

File: backend/app/routes/admin/changePassword.js [hunk 1]
```diff
@@ -6,19 +6,32 @@ const constants = require('../../../constants');
...
       statusCode: 400,
       message: 'Wrong old password',
       user: userRecord.email,
-      errStack: 'Auth Login',
+      errStack: 'Password change',
+    });
+    return next(error);
+  }
+
+  if (oldPassword === newPassword) {
```
Comment (jackfrost13) [inline backend/app/routes/admin/changePassword.js:29]: this should be the first thing before making DB call to save on latency

File: backend/app/routes/admin/changePassword.js [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+      user: userRecord.email,
+      errStack: 'Password change',
+    });
+    return next(error);
+  }
+
+  const hashedPassword = await argon2.hash(newPassword);
+
+  const [err, user] = await to(
+    Admin.findOne({ email: userRecord.email }, () => {
```
Comment (jackfrost13) [inline backend/app/routes/admin/changePassword.js:42]: you should not use findOne and then save method. You know which fields to update, so you should use find and update method. It saves time eventually resulting in better latency

File: backend/app/routes/admin/changePassword.js [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+      statusCode: 500,
+      message: 'Mongo Error: Insertion Failed',
+      errStack: err,
+      user: userRecord.email,
+    });
+    return next(error);
+  }
+
+  //   eslint-disable-next-line no-underscore-dangle
+  const response = { ...user._doc };
```
Comment (jackfrost13) [inline backend/app/routes/admin/changePassword.js:59]: we do not require to return the entire user data on password change. Just the email id should be fine

File: backend/app/routes/admin/changePassword.js [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+      user: userRecord.email,
+      errStack: 'Password change',
+    });
+    return next(error);
+  }
+
+  if (oldPassword === newPassword) {
+    const error = new ErrorHandler(constants.ERRORS.INPUT, {
+      statusCode: 400,
+      message: 'Old password cannot be same as new password.',
```
Comment (jackfrost13) [inline backend/app/routes/admin/changePassword.js:32]: `new password cannot be identical to old password`

**Review by jackfrost13 — COMMENTED** [review]
@N00Bmaster2810 can you please incorporate these changes and open a new PR? apologies, I couldn't review it earlier, had a busy week

File: backend/app/routes/admin/changePassword.js (hunk 4 above)
Comment (N00Bmaster2810) [reply backend/app/routes/admin/changePassword.js:32]: @jackfrost13 so this check is not required?

Comment (jackfrost13) [reply backend/app/routes/admin/changePassword.js:32]: I meant the message
`message: 'Old password cannot be same as new password.' `to
`message: 'new password cannot be identical to old password'`

Comment (N00Bmaster2810) [reply backend/app/routes/admin/changePassword.js:32]: ohh got it thank you

File: backend/app/routes/admin/changePassword.js (hunk 1 above)
Comment (N00Bmaster2810) [reply backend/app/routes/admin/changePassword.js:29]: @jackfrost13 I didn't understand?

Comment (jackfrost13) [reply backend/app/routes/admin/changePassword.js:29]: You should check this before making any DB call

Comment (N00Bmaster2810) [reply backend/app/routes/admin/changePassword.js:29]: ok . so @jackfrost13 after making these changes should I open a new PR?

Comment (jackfrost13) [reply backend/app/routes/admin/changePassword.js:29]: yes, please change api testing also. Then I can review it again if required

Comment (N00Bmaster2810) [reply backend/app/routes/admin/changePassword.js:29]: ok

**Pull request merged**